### DISCUSSION TITLE
Diversification of Unit Integrals

### DISCUSSION THREAD

**Discussion by Naktakala** [chi-tech/chi-tech#173] [Ideas]
## What we currently have
Currently we compute all of the following unit integrals
- `IntV_gradShapeI_gradShapeJ`
- `IntV_shapeI_gradshapeJ`
- `IntV_shapeI_shapeJ`
- `IntV_shapeI`
- `IntV_gradshapeI`
- `IntS_shapeI_shapeJ`
- `IntS_shapeI`
- `IntS_shapeI_gradshapeJ`, and auxiliary-info
- `FaceDofMapping`

Transport, Diffusion and Monte Carlo simulations all overlap with the mass-matrices and maybe some other matrices but in many cases these simulations don’t require the precomputation of all these matrices (it takes an enormous amount of memory making 200k cell diffusion simulations on a single processor consume many Gigs of memory).

We control the precomputation of these integrals with a flag `COMPUTE_UNIT_INTEGRALS`.

## What we need
First we need to move `FaceDofMapping` out of the unit integrals. Then we need the ability to compute these items separately with these possible flags:
- `PRECOMPUTE_INTV_GRADSHAPEI_GRADSHAPEJ`
- `PRECOMPUTE_INTV_SHAPEI_GRADSHAPEJ`
- `PRECOMPUTE_INTV_SHAPEI_SHAPEJ`
- `PRECOMPUTE_INTV_SHAPEI`
- `PRECOMPUTE_INTV_GRADSHAPEI`
- `PRECOMPUTE_INTS_SHAPEI_SHAPEJ`
- `PRECOMPUTE_INTS_SHAPEI`
- `PRECOMPUTE_INTS_SHAPEI_GRADSHAPEJ`

**Comment by Naktakala** [comment]
A rough depiction of how `SpatialDiscretization` methods are currently operating:

Note: The IntV or IntS quantities denotes cell FE matrices

![image](https://user-images.githubusercontent.com/25042529/127701841-3d867ab9-6d72-4c05-91bd-10b2a698799a.png)

**Comment by Naktakala** [comment]
As my experience grew I now know that the `SpatialDiscretization` determines the nodal-structure. Every solver can slap a DOF-structure onto this nodal-structure to define matrices and vectors of data. I really want the `SpatialDiscretization` methods to work as:
![image](https://user-images.githubusercontent.com/25042529/127703352-2ad9fb73-7155-4777-862d-59a2fd953897.png)

Here a `SpatialDiscretization` is a stand-alone entity that can be asked to store a few things (CellMappings, specific Cell-matrices, quadrature point data etc.) and these things are shareable amongst processes.

**Comment by Naktakala** [comment]
Design requirements (we will use `IntV_shapeI_shapeJ` to refer to a generic cell FE matrix):
1. To get a cell FE matrix (i.e., `IntV_shapeI_shapeJ`), a call should be made to the spatial discretization to `GetIntV_shapeI_shapeJ(cell)`
2. A `SpatialDiscretization` should be able to do either:
  a) a fast call to `GetIntV_shapeI_shapeJ()` to get a precomputed and stored `IntV_shapeI_shapeJ`, or 
  b) a slow call to `GetIntV_shapeI_shapeJ()` to compute and return `IntV_shapeI_shapeJ`

_Context for req. 2a: A solver should be able to call something like `PreComputeCellFEMatrices(which_ones)` where the `which_ones` flags indicate which cell FE matrices to precompute and store. This will incur storage, however, clients of this capability (like transport solvers) can then speed up because they can reuse these matrices._

_Context for req. 2b: Not all solvers require a lot of DOFs per node, and many solvers access the cell matrices either once per matrix assembly, or as a once-off post process step. In this context we can consider a simple 200k diffusion simulation which should consume ~200 Mb at max. If one precomputes the cell FE matrices then this balloons to a couple of Gigabytes without any increase in speed (matrix assembly happens once). In this regard the code should be able to get a cell matrix on the fly._

**Reply by wdhawkins** [reply]
With these requirements in mind, I can replace PR#190 with something much simpler that still accomplishes many of the same goals - namely, the simplification and eventual removal of UnitIntegralData.  The new PR simply does away with the scratch UnitIntegral and QP objects and always returns a temporary in the case that the integral/qp data has not been initialized. While the initial implementation will incur additional overhead because we populate an entire UnitIntegralData object, future implementations (corresponding to 2b) won't have this problem since they should be ability to request and return single matrices.

**Reply by dominiccaron** [reply]
I believe that the overhead of the short-term implementation that you mention will not differ greatly from the current implementation, with the exception of a little more cost to create and to destroy the temporary on-the-fly. In any case, as you make clear, this is only short-term. I support the proposed approach.
